### PR TITLE
Desktop Analytics

### PR DESCRIPTION
**Pull request by neuroscr**
Address oxen-io/session-desktop-temp#1081 

Meh:
* New Conversation Started (+1x Conversation Opened)

Verified 100%
* Group Message Sent
* RSS Feed Opened
* Loki Public Chat Opened
* Conversation Opened (i.e. a private chat)
* Sent Message Using Swarm API
* Seed Restored

Less tested
+ Seed Created
+ Seed Node Failed (i.e. got an error response when trying to update the random snode pool)
+ Unreachable Snode
+ Migrated Snode
+ Failed to Send Group Message (after having retried)
+ Failed to Send Message Using Swarm API (after having retried)

### CODE REVIEW

File: libtextsecure/account_manager.js [hunk 1]
```diff
@@ -27,6 +27,8 @@
   function AccountManager(username, password) {
     // this.server = window.WebAPI.connect({ username, password });
     this.pending = Promise.resolve();
+    // set up mixpanel
+    window.mixpanel = new window.LokiMixpanelAPI();
```
Comment (Mikunj) [inline libtextsecure/account_manager.js]: Might be better moving this up below `window.textsecure = window.textsecure || {};`.

`window.mixpanel = window.mixpanel || new window.LokiMixpanelAPI()`

thoughts?

Comment (neuroscr) [reply libtextsecure/account_manager.js]: 99% sure it's not created in account_manager.js at that point but can't hurt. Applied associated changes.

**Review by Mikunj — APPROVED** [review]
Looks good :)

File: js/modules/loki_mixpanel.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+/* eslint-disable class-methods-use-this */
```
Comment (sachaaaaa) [inline js/modules/loki_mixpanel.js]: stale ?

File: preload.js [hunk 3]
```diff
@@ -328,6 +328,8 @@ window.LokiPublicChatAPI = require('./js/modules/loki_public_chat_api');
 
 window.LokiRssAPI = require('./js/modules/loki_rss_api');
 
+window.LokiMixpanelAPI = require('./js/modules/loki_mixpanel.js');
```
Comment (sachaaaaa) [inline preload.js]: You can actually instantiate `window.mixpanel` just the one time here.
`preload.js` gets executed before the front-end kicks in, so if you do
```js
const LokiMixpanelAPI = require('./js/modules/loki_mixpanel.js');
window.mixpanel = new LokiMixPanelAPI();
```
here, you should be able to use it everywhere else (even in modules! the first comments tells the linter what global variable we are going to access)

File: js/modules/loki_snode_api.js [hunk 4]
```diff
@@ -6,6 +6,9 @@ const dns = require('dns');
 const process = require('process');
 const { rpc } = require('./loki_rpc');
 const natUpnp = require('nat-upnp');
+const LokiMixpanelAPI = require('./loki_mixpanel.js');
+
+const Mixpanel = new LokiMixpanelAPI();
```
Comment (sachaaaaa) [inline js/modules/loki_snode_api.js]: Just use `window.mixpanel` directly in the code below

File: js/modules/loki_public_chat_api.js [hunk 5]
```diff
@@ -3,6 +3,9 @@ clearTimeout, MessageController */
 const EventEmitter = require('events');
 const nodeFetch = require('node-fetch');
 const { URL, URLSearchParams } = require('url');
+const LokiMixpanelAPI = require('./loki_mixpanel.js');
+
+const Mixpanel = new LokiMixpanelAPI();
```
Comment (sachaaaaa) [inline js/modules/loki_public_chat_api.js]: Same, just use `window.mixpanel` directly below

File: js/modules/loki_mixpanel.js [hunk 6]
```diff
@@ -0,0 +1,17 @@
...
+
+const EventEmitter = require('events');
+const Mixpanel = require('mixpanel');
+// require('setimmediate');
+
+class LokiMixpanelAPI extends EventEmitter {
```
Comment (sachaaaaa) [inline js/modules/loki_mixpanel.js]: Is there a need for deriving EventEmitter?

File: js/modules/loki_message_api.js [hunk 7]
```diff
@@ -4,6 +4,9 @@
 
 const _ = require('lodash');
 const { rpc } = require('./loki_rpc');
+const LokiMixpanelAPI = require('./loki_mixpanel.js');
+
+const Mixpanel = new LokiMixpanelAPI();
```
Comment (sachaaaaa) [inline js/modules/loki_message_api.js]: Same, just use `window.mixpanel` directly

File: js/background.js [hunk 8]
```diff
@@ -230,6 +230,9 @@
     window.feeds = [];
     window.lokiMessageAPI = new window.LokiMessageAPI(ourKey);
     window.lokiPublicChatAPI = new window.LokiPublicChatAPI(ourKey);
+    // set up mixpanel
+    window.mixpanel = window.mixpanel || new window.LokiMixpanelAPI(); // incase account manager already set it up
```
Comment (sachaaaaa) [inline js/background.js]: remove if you already set it up in `preload.js` ;)

File: js/modules/loki_mixpanel.js (hunk 2 above)
Comment (neuroscr) [reply js/modules/loki_mixpanel.js]: addressed

File: preload.js (hunk 3 above)
Comment (neuroscr) [reply preload.js]: addressed

File: js/modules/loki_snode_api.js (hunk 4 above)
Comment (neuroscr) [reply js/modules/loki_snode_api.js]: addressed

File: js/modules/loki_public_chat_api.js (hunk 5 above)
Comment (neuroscr) [reply js/modules/loki_public_chat_api.js]: addressed

File: js/modules/loki_mixpanel.js (hunk 6 above)
Comment (neuroscr) [reply js/modules/loki_mixpanel.js]: nope, removed.

File: js/modules/loki_message_api.js (hunk 7 above)
Comment (neuroscr) [reply js/modules/loki_message_api.js]: addressed

File: js/background.js (hunk 8 above)
Comment (neuroscr) [reply js/background.js]: removed.

**Comment by neuroscr** [comment]
Looks like `New Conversation Started` broke with the changes

**Review by sachaaaaa — APPROVED** [review]
Looks good!

**Pull request merged**